packages/opencensus-instrumentation-http2/src/http2.ts

```diff
@@ -0,0 +1,283 @@
+
+    shimmer.wrap(moduleExports, 'connect', this.patchConnect());
+
+    return moduleExports;
+  }
+
+  /** Unpatches all HTTP2 patched function. */
+  applyUnpatch(): void {
+    if (this.patchedCreateServer) {
+      shimmer.unwrap(plugin.server.constructor.prototype, 'emit');
+    } else {
+      shimmer.unwrap(plugin.moduleExports, 'createServer');
+      shimmer.unwrap(plugin.moduleExports, 'createSecureServer');
+    }
+
+    if (this.patchedConnect) {
+      shimmer.unwrap(plugin.client, 'request');
+    } else {
+      shimmer.unwrap(plugin.moduleExports, 'connect');
+    }
+  }
+
+  patchConnect() {
+    const plugin = this;
+    return (original: ConnectFunction):
+               types.Func<http2.ClientHttp2Session> => {
+      return function patchedConnect(authority: string):
+          http2.ClientHttp2Session {
+            plugin.client = original.apply(this, arguments);
+            shimmer.wrap(plugin.client, 'request', plugin.patchRequest());
+
+            shimmer.unwrap(plugin.moduleExports, 'connect');
+            plugin.patchedConnect = true;
+            plugin.clientUrl = url.parse(authority);
+
+            return plugin.client;
+          };
+    };
+  }
+
+  patchRequest() {
+    const plugin = this;
+    return (original: RequestFunction): types.Func<http2.ClientHttp2Stream> => {
+      return function patchedRequest(
+                 headers: http2.IncomingHttpHeaders): http2.ClientHttp2Stream {
+        // Do not trace ourselves
+        if (headers['x-opencensus-outgoing-request']) {
+          return original.apply(this, arguments);
+        }
+
+        const request = original.apply(this, arguments);
+        plugin.tracer.wrapEmitter(request);
+
+        const traceOptions = {
+          name: `${headers[':method'] ? headers[':method'] : 'GET'} ${
+              headers[':path']}`,
+          type: 'CLIENT',
+        };
+
+        // Checks if this outgoing request is part of an operation by checking
+        // if there is a current root span, if so, we create a child span. In
+        // case there is no root span, this means that the outgoing request is
+        // the first operation, therefore we create a root span.
+        if (!plugin.tracer.currentRootSpan) {
+          return plugin.tracer.startRootSpan(
+              traceOptions,
+              plugin.makeHttp2RequestTrace(request, headers, plugin));
+        } else {
+          const span = plugin.tracer.startChildSpan(
+              traceOptions.name, traceOptions.type);
+          return (plugin.makeHttp2RequestTrace(request, headers, plugin))(span);
+        }
+      };
+    };
+  }
+
+  makeHttp2RequestTrace(
+      request: http2.ClientHttp2Stream, headers: http2.IncomingHttpHeaders,
+      plugin: Http2Plugin): types.Func<http2.ClientHttp2Stream> {
+    return (span: types.Span): http2.ClientHttp2Stream => {
+      if (!span) return request;
+
+      const setter = {
```

Comment:
Does `const setter: types.HeaderSetter = ...` work here? Prefer that to typecasting.